extender.go

```diff
@@ -113,38 +121,272 @@ func (f *ExtenderFactory) verifyAPIs(groupExt []buildpack.GroupElement, logger l
-		buildDockerfile, err := e.buildDockerfileFor(ext.ID)
+func (e *Extender) Extend(kind string, logger log.Logger) error {
+	switch kind {
+	case buildpack.DockerfileKindBuild:
+		return e.extendBuild(logger)
+	case buildpack.DockerfileKindRun:
+		return e.extendRun(logger)
+	default:
+		return nil
+	}
+}
+
+func (e *Extender) extendBuild(logger log.Logger) error {
+	origBaseImage, err := e.DockerfileApplier.ImageFor(e.ImageRef)
+	if err != nil {
+		return fmt.Errorf("getting build image to extend: %w", err)
+	}
+
+	extendedImage, err := e.extend(buildpack.DockerfileKindBuild, origBaseImage, logger)
+	if err != nil {
+		return fmt.Errorf("extending build image: %w", err)
+	}
+
+	if err = setImageEnvVarsInCurrentContext(extendedImage); err != nil {
+		return fmt.Errorf("setting environment variables from extended image in current context: %w", err)
+	}
+
+	return e.DockerfileApplier.Cleanup()
+}
+
+func setImageEnvVarsInCurrentContext(image v1.Image) error {
+	configFile, err := image.ConfigFile()
+	if err != nil || configFile == nil {
+		return fmt.Errorf("getting config for extended image: %w", err)
+	}
+	for _, env := range configFile.Config.Env {
+		parts := strings.Split(env, "=")
+		if len(parts) != 2 {
+			return fmt.Errorf("parsing env '%s': expected format 'key=value'", env)
+		}
+		if err := os.Setenv(parts[0], parts[1]); err != nil {
+			return fmt.Errorf("setting env: %w", err)
+		}
+	}
+	return nil
+}
+
+func (e *Extender) extendRun(logger log.Logger) error {
+	origBaseImage, err := e.DockerfileApplier.ImageFor(e.ImageRef)
+	if err != nil {
+		return fmt.Errorf("getting run image to extend: %w", err)
+	}
+
+	origTopLayer, err := topLayer(origBaseImage)
+	if err != nil {
+		return fmt.Errorf("getting original run image top layer: %w", err)
+	}
+
+	extendedImage, err := e.extend(buildpack.DockerfileKindRun, origBaseImage, logger)
+	if err != nil {
+		return fmt.Errorf("extending run image: %w", err)
+	}
+
+	if err = e.saveSelective(extendedImage, origTopLayer); err != nil {
+		return fmt.Errorf("copying selective image to output directory: %w", err)
+	}
+
+	return e.DockerfileApplier.Cleanup()
+}
+
+func topLayer(image v1.Image) (string, error) {
+	manifest, err := image.Manifest()
+	if err != nil {
+		return "", fmt.Errorf("getting image manifest: %w", err)
+	}
+	layers := manifest.Layers
+	if len(layers) == 0 {
+		return "", nil
+	}
+	layer := layers[len(layers)-1]
+	return layer.Digest.String(), nil
+}
+
+func (e *Extender) saveSelective(image v1.Image, origTopLayerHash string) error {
+	// save sparse image (manifest and config)
+	imageHash, err := image.Digest()
+	if err != nil {
+		return fmt.Errorf("getting image hash: %w", err)
+	}
+	toPath := filepath.Join(e.ExtendedDir, "run", imageHash.String())
+	layoutPath, err := selective.Write(toPath, empty.Index) // FIXME: this should use the imgutil layout/sparse package instead, but for some reason sparse.NewImage().Save() fails when the provided base image is already sparse
```

Comment:
This was a bummer to discover. When we update imgutil we can remove some logic (e.g., `topLayer`) from this file.